### PR TITLE
Bugs Fixed

### PR DESCRIPTION
**Pull request by md-kasim**
Use of CKEDITOR by edit answer,  add comment,
option for edit comment,
option for uploading image in fosscategory,
sorting of rows of table in search.html
and many more

### CODE REVIEW

**Comment by md-kasim** [comment]
Use of CKEDITOR by edit answer, add comment,
option for edit comment,
option for uploading image in fosscategory,
sorting of rows of table in search.html
and many more

File: forums/wsgi.py [hunk 1]
```diff
@@ -28,6 +28,8 @@
 # os.environ["DJANGO_SETTINGS_MODULE"] = "forums.settings"
 os.environ["DJANGO_SETTINGS_MODULE"] = "forums.settings"
 
+# os.environ['HTTPS'] = "on"
```
Comment (CruiseDevice) [inline forums/wsgi.py]: Remove commented code.

File: static/website/css/main.css [hunk 2]
```diff
@@ -98,7 +98,7 @@ body {
 /* } */
 #content .question {
     position: relative;
-    padding: 30px 0 15px 0;
+    /*padding: 30px 0 15px 0;*/
```
Comment (prathamesh920) [inline static/website/css/main.css]: Remove commented code.
From other places as well.

File: website/admin.py [hunk 3]
```diff
@@ -39,3 +46,6 @@ class GroupAdmin(BaseGroupAdmin):
 admin.site.register(Answer, AnswerAdmin)
 admin.site.register(AnswerComment)
 admin.site.register(FossCategory)
+
+
```
Comment (prathamesh920) [inline website/admin.py]: Why 2 blank lines?

File: website/cron.py [hunk 4]
```diff
@@ -23,36 +23,36 @@ def unanswered_notification(self):
         import datetime as DT
 
         try:
-            weekago = DT.date.today() - DT.timedelta(days = 6)
+            weekago = DT.date.today() - DT.timedelta(days=6)
             questions = Question.objects\
-                        .filter(date_created__lte = weekago, is_spam = 0)
+                .filter(date_created__lte=weekago, is_spam=0)
         except Exception as e:
```
Comment (prathamesh920) [inline website/cron.py:29]: Just a suggestion below:

What Exception is caught?
Also, filter is used so if no match then it will return empty set. But in exception it is "No questions found".
Also, print can be removed.

File: website/templatetags/permission_tags.py [hunk 5]
```diff
@@ -2,22 +2,57 @@
...
             return False
-    except:
+    except BaseException:
         return False
+
+
 register.filter(is_moderator)
+
+
+def comment_id(answer):
```
Comment (prathamesh920) [inline website/templatetags/permission_tags.py:32]: This is not clear.
Can u explain?

File: website/templatetags/permission_tags.py [hunk 6]
```diff
@@ -2,22 +2,57 @@
...
+    return "comment" + str(answer.id)
+
+
+register.filter(comment_id)
+
+
+def havenot_comments(answer):
+    comment = answer.answercomment_set.filter(is_active=True)
```
Comment (prathamesh920) [inline website/templatetags/permission_tags.py]: Suggestion:
```return answer.answercomment_set.filter(is_active=True).exists()```

Also, function name can be better and consistent like underscores between the words.

File: website/views.py [hunk 7]
```diff
@@ -17,27 +17,52 @@
...
-categories = FossCategory.objects.order_by('name')
 
 # Function to check if user is in any moderator group
-def is_moderator(user):
-    return user.groups.count() > 0
 
-# Function to check if user is anonymous, and if not he/she has first_name and last_name
+
+
+def is_moderator(user,question=None):
```
Comment (prathamesh920) [inline website/views.py]: space after comma
(PEP8)

File: website/views.py [hunk 8]
```diff
@@ -17,27 +17,52 @@
...
+    if question:
+        if user.groups.filter(moderatorgroup=ModeratorGroup.objects.get(category=question.category)).exists():
+            return True
+    elif user.groups.count() > 0:
```
Comment (prathamesh920) [inline website/views.py]: Why only ```user.groups.count() > 0``` is not enough to say user is moderator?
Why ```user.groups.filter(moderatorgroup=.....``` is needed?

File: website/views.py [hunk 9]
```diff
@@ -17,27 +17,52 @@
...
 
 # for home page
-def home(request):
 
-    settings.MODERATOR_ACTIVATED = False
 
-    questions = Question.objects.all().order_by('-date_created').filter(is_spam = False)
+def home(request):
+    settings.MODERATOR_ACTIVATED = False
+    next = request.GET.get('next', '')
```
Comment (prathamesh920) [inline website/views.py:49]: ```next``` is a builtin in python, so be careful about the namespace issue.

File: website/views.py [hunk 10]
```diff
@@ -46,41 +71,57 @@ def home(request):
...
         'questions': questions,
     }
     return render(request, 'website/templates/questions.html', context)
 
 # get particular question, with votes,anwsers
-def get_question(request, question_id = None, pretty_url = None):
 
-    question = get_object_or_404(Question, id = question_id)
+
+def get_question(request, question_id=None, pretty_url=None):
```
Comment (prathamesh920) [inline website/views.py:88]: This function is not clear.
Kindly explain.

File: website/views.py [hunk 11]
```diff
@@ -109,15 +150,16 @@ def get_question(request, question_id = None, pretty_url = None):
...
-                  .filter(question_id = question_id).distinct()
-
+    question = get_object_or_404(Question, id=question_id, is_active=True)
     if (request.method == 'POST'):
 
         form = AnswerQuestionForm(request.POST, request.FILES)
         answer = Answer()
         answer.uid = request.user.id
+        if request.POST['body'].strip() == "":
+            messages.error(request,"Answer cann't be empty or only blank spaces.")
```
Comment (prathamesh920) [inline website/views.py]: Typo

File: website/views.py [hunk 12]
```diff
@@ -164,21 +209,44 @@ def question_answer(request, question_id):
...
-                headers = {"Content-type":"text/html;charset=iso-8859-1"}
+                headers={"Content-type": "text/html;charset=iso-8859-1"}
             )
             email.attach_alternative(message, "text/html")
             email.content_subtype = 'html'  # Main content is text/html
             email.mixed_subtype = 'related'
-            email.send(fail_silently = True)
+            email.send(fail_silently=True)
+
+            answer_sets = Answer.objects.filter(
```
Comment (prathamesh920) [inline website/views.py]: variable name can be ```answers``` instead of answer_sets.
Similarly for the below defined variables.

File: website/views.py [hunk 13]
```diff
@@ -164,21 +209,44 @@ def question_answer(request, question_id):
...
+                question_id=question_id, is_active=True).distinct()
+            comment_set = []
+            mail_ids = []
+            for answer_id in answer_sets:
+                comment_set.append(
+                    AnswerComment.objects.values('uid').filter(
+                        answer=answer_id, is_active=True).distinct())
+                mail_ids.append(answer_id.uid)
+            for x in comment_set:
```
Comment (prathamesh920) [inline website/views.py]: instead of ```x``` use readable name like ```comment```

File: website/views.py [hunk 14]
```diff
@@ -164,21 +209,44 @@ def question_answer(request, question_id):
...
+                for y in x:
```
Comment (prathamesh920) [inline website/views.py]: ```y``` is also not clear!

File: website/views.py [hunk 15]
```diff
@@ -164,21 +209,44 @@ def question_answer(request, question_id):
...
+                    mail_ids.append(y['uid'])
+            mail_ids = set(mail_ids)
+            mail_ids.difference_update({question.user.id, request.user.id})
+
+            # Notification for all user in this thread
+            for x in mail_ids:
```
Comment (prathamesh920) [inline website/views.py]: Kindly give proper variable names.

File: static/website/css/main.css (hunk 2 above)
Comment (md-kasim) [reply static/website/css/main.css]: comments removed

File: website/admin.py (hunk 3 above)
Comment (md-kasim) [reply website/admin.py]: Blank lines removed

File: website/views.py [hunk 16]
```diff
@@ -292,126 +373,189 @@ def answer_comment(request):
...
+                    sender_email, to,
+                    bcc=[bcc_email],
+                    headers={"Content-type": "text/html;charset=iso-8859-1"}
+                )
+                email.attach_alternative(message, "text/html")
+                email.content_subtype = 'html'  # Main content is text/html
+                email.mixed_subtype = 'related'
+                email.send(fail_silently=True)
+
+            comment_set = []
```
Comment (prathamesh920) [inline website/views.py]: The below code looks similar to the code on the top.
So can we have a function for this so that the redundancy is reduced.

File: website/views.py [hunk 17]
```diff
@@ -570,44 +753,59 @@ def edit_question(request, question_id):
...
+            answer_sets = Answer.objects.filter(
+                question_id=question_id, is_active=True).distinct()
+            comment_set = []
+            mail_ids = [question.user.id]
+            for answer in answer_sets:
+                comment_set.append(
+                    AnswerComment.objects.values('uid').filter(
+                        answer=answer, is_active=True).distinct())
+                mail_ids.append(answer.uid)
+            for x in comment_set:
```
Comment (prathamesh920) [inline website/views.py]: Again similar code.

File: website/views.py [hunk 18]
```diff
@@ -622,40 +820,169 @@ def edit_question(request, question_id):
...
```
Comment (prathamesh920) [inline website/views.py]: Again similar code

File: website/views.py [hunk 19]
```diff
@@ -622,40 +820,169 @@ def edit_question(request, question_id):
...
+        answer_sets = Answer.objects.filter(
+            question_id=question_id, is_active=True).distinct()
+        comment_set = []
+        mail_ids = [answer.question.user.id]
+        for answer in answer_sets:
+            comment_set.append(
+                AnswerComment.objects.values('uid').filter(
+                    answer=answer, is_active=True).distinct())
+            mail_ids.append(answer.uid)
+        for x in comment_set:
```
Comment (prathamesh920) [inline website/views.py]: Again similar.

File: website/views.py [hunk 20]
```diff
@@ -998,91 +1374,279 @@ def ajax_tutorials(request):
...
+                answer_sets = Answer.objects.filter(
+                    question_id=question_id, is_active=True).distinct()
+                comment_set = []
+                mail_ids = [answer.question.user.id]
+                for ans in answer_sets:
+                    comment_set.append(
+                        AnswerComment.objects.values('uid').filter(
+                            answer=ans, is_active=True).distinct())
+                    mail_ids.append(ans.uid)
+                for x in comment_set:
```
Comment (prathamesh920) [inline website/views.py]: Similar!

File: website/views.py [hunk 21]
```diff
@@ -998,91 +1374,279 @@ def ajax_tutorials(request):
...
+                answer_sets = Answer.objects.filter(
+                    question_id=question_id, is_active=True).distinct()
+                comment_set = []
+                mail_ids = [comment.answer.question.user.id]
+                for ans in answer_sets:
+                    comment_set.append(
+                        AnswerComment.objects.values('uid').filter(
+                            answer=ans, is_active=True).distinct())
+                    mail_ids.append(ans.uid)
+                for x in comment_set:
```
Comment (prathamesh920) [inline website/views.py]: Similar!

File: website/views.py [hunk 22]
```diff
@@ -998,91 +1374,279 @@ def ajax_tutorials(request):
...
+
 
+@csrf_exempt
+def ajax_answer_comment_update(request):
+    if request.method == 'POST':
+        cid = request.POST['comment_id']
+        body = request.POST['comment_body']
+        try:
+            comment = get_object_or_404(AnswerComment, pk=cid, is_active=True)
+        except BaseException:
```
Comment (prathamesh920) [inline website/views.py]: Here ```try ... except ``` is not needed.

File: website/views.py [hunk 23]
```diff
@@ -998,91 +1374,279 @@ def ajax_tutorials(request):
...
+                answer_sets = Answer.objects.filter(
+                    question_id=question_id, is_active=True).distinct()
+                comment_set = []
+                mail_ids = [comment.answer.question.user.id]
+                for ans in answer_sets:
+                    comment_set.append(
+                        AnswerComment.objects.values('uid').filter(
+                            answer=ans, is_active=True).distinct())
+                    mail_ids.append(ans.uid)
+                for x in comment_set:
```
Comment (prathamesh920) [inline website/views.py]: Similar!

**Review by prathamesh920 — CHANGES_REQUESTED** [review]
Kindly make the requested changes.
Also, it will be good to have a custom css file instead of inline CSS.

File: static/website/js/thread-user.js [hunk 24]
```diff
@@ -1,327 +1,71 @@
...
     $answer_edit.click(function(e) {
         var target = $(this).data("target");
-        answerNicEditor.addInstance(target);
-        $(this).parents("div.answer").prepend($answerPanelWrapper);
-        $answerPanelWrapper.show();
-        $('#answerPanelWrapper .nicEdit-panelContain').parent().width('100%');
-        $('#answerPanelWrapper .nicEdit-panelContain').parent().next().width('100%');
+        var id = $(this).attr("id");
+        $("#"+id+"1").show();
+        $("#"+id+"2").hide();
```
Comment (prathamesh920) [inline static/website/js/thread-user.js:13]: I am some what not convinced with these lines of code.
@prashantsinalkar kindly look into this JS

File: website/templatetags/permission_tags.py (hunk 5 above)
Comment (md-kasim) [reply website/templatetags/permission_tags.py:32]: In get-question.html
line no. 355
{% render_field form.body class+="form-control body" id=answer|comment_id %}
comment_id function is being used to give an id to the CKEditor for posting the comment.
because id=comment{{ answer.id }} is not working.

File: website/templatetags/permission_tags.py (hunk 6 above)
Comment (md-kasim) [reply website/templatetags/permission_tags.py]: done

File: website/views.py (hunk 7 above)
Comment (md-kasim) [reply website/views.py]: done

File: website/views.py (hunk 8 above)
Comment (md-kasim) [reply website/views.py]: while editing the question or its answer or comments, moderator should be checked whether he is belong to the moderator group of the question's category or not
Otherwise User of any moderator group of specific category can edit the questions of any other categories

File: website/views.py (hunk 10 above)
Comment (md-kasim) [reply website/views.py:88]: this is being used to show the questions and its answers and comments in user panel or moderator panel.

File: website/views.py (hunk 22 above)
Comment (md-kasim) [reply website/views.py]: done

File: website/views.py (hunk 14 above)
Comment (md-kasim) [reply website/views.py]: solved

File: website/views.py (hunk 13 above)
Comment (md-kasim) [reply website/views.py]: solved

File: website/views.py (hunk 15 above)
Comment (md-kasim) [reply website/views.py]: issue solved

File: website/views.py (hunk 16 above)
Comment (md-kasim) [reply website/views.py]: functions has been defined
issue solved

File: website/views.py (hunk 17 above)
Comment (md-kasim) [reply website/views.py]: issue solved

File: website/views.py (hunk 18 above)
Comment (md-kasim) [reply website/views.py]: issue solved

File: website/views.py (hunk 19 above)
Comment (md-kasim) [reply website/views.py]: issue solved

File: website/views.py (hunk 20 above)
Comment (md-kasim) [reply website/views.py]: issue solved

File: website/views.py (hunk 21 above)
Comment (md-kasim) [reply website/views.py]: issue solved

File: website/views.py (hunk 23 above)
Comment (md-kasim) [reply website/views.py]: issue solved

File: website/views.py (hunk 12 above)
Comment (md-kasim) [reply website/views.py]: variable has been changed with answers

File: website/views.py (hunk 11 above)
Comment (md-kasim) [reply website/views.py]: removed

File: static/website/templates/get-question.html [hunk 25]
```diff
@@ -149,10 +150,26 @@ <h4><u>Answers:</u></h4>
...
+                <form id="form{{ answer.id }}" method="POST"  action="{% url 'website:ajax_answer_update' %}">
+                    {% csrf_token %}
+                    <input type="hidden" name="answer_id" value="{{ answer.id }}">
+                <div class="edit-answer" id="ans{{ answer.id }}1">
+                    {% render_field form.body class+="form-control body" name="answer_body" id=answer.id %}
+                </div>
+                </form>
+                {% endif %}
+                <div id="ans{{ answer.id }}2">{{ answer.body| safe }}</div>
+                {% if user|is_moderator and MODERATOR_ACTIVATED or user|can_edit:answer and answer|havenot_comments %}
```
Comment (sunilshetye) [inline static/website/templates/get-question.html]: use brackets here to make logic clear. use of **and** as well as **or** in the same conditional without brackets is not recommended.

File: static/website/templates/get-question.html [hunk 26]
```diff
@@ -194,15 +211,20 @@ <h4><u>Answers:</u></h4>
                 </span>
             </span>
 
-            {% if user|is_moderator and MODERATOR_ACTIVATED  %}
+            {% if user|is_moderator and MODERATOR_ACTIVATED or user|can_edit:answer and answer|havenot_comments %}
```
Comment (sunilshetye) [inline static/website/templates/get-question.html]: use brackets

File: static/website/templates/moderator/index.html [hunk 27]
```diff
@@ -11,9 +11,16 @@
...
-          $("#myTable").DataTable({
+          var table = $("#myTable").DataTable({
           // 'sPaginationType': "scrolling"
           });
+          $('#category').change(function(e){
+            var selectedCategory = $(this).children("option:selected").val();
+            if( selectedCategory != "All Categories")
+                table.column(1).search('^' + selectedCategory + '$', true, false, true).draw();
+            else
+                table.column(1).search('').draw();
```
Comment (sunilshetye) [inline static/website/templates/moderator/index.html:22]: is it necessary to search an empty string here?

File: forums/settings.py [hunk 28]
```diff
@@ -164,6 +164,8 @@
     'ckeditor_uploader',
 )
 
+SESSION_EXPIRE_AT_BROWSER_CLOSE = True
```
Comment (sunilshetye) [inline forums/settings.py:167]: add explanation for this setting.

File: static/website/css/main.css [hunk 29]
```diff
@@ -496,22 +491,27 @@ table .question a{
...
 }
 
-.comment:hover a.delete-comment, a.delete-comment:hover {
+.comment:hover a.delete-comment, a.delete-comment:hover, .comment:hover a.restore-comment, a.restore-comment:hover, .comment:hover a.edit-comment, a.edit-comment:hover, .comment:hover a.save-comment, a.save-comment:hover {
     display: inline-block;
 }
+
+.saved {
+    color: black;
+    /*display: none;*/
```
Comment (sunilshetye) [inline static/website/css/main.css]: remove comments

File: static/website/templates/get-question.html [hunk 30]
```diff
@@ -91,8 +97,14 @@
 
         <span class="modify">
             {% if user|can_edit:question and not ans_count or user|is_moderator and MODERATOR_ACTIVATED %}
```
Comment (sunilshetye) [inline static/website/templates/get-question.html]: use brackets here also (though not part of this change)

File: static/website/templates/get-question.html [hunk 31]
```diff
@@ -149,10 +150,26 @@ <h4><u>Answers:</u></h4>
             </div>
         </div>
 
-        <div class="answer {% if user|is_moderator %}editable{% endif %}" id="answer{{answer.id}}" style="margin-left: 63px;">                               
-            
+        <div class="answer {% if user|is_moderator and MODERATOR_ACTIVATED or user|can_edit:answer %}editable{% endif %}" id="answer{{answer.id}}" style="margin-left: 63px;">                               
             <div class="body" id="body{{ answer.id }}" style="padding-bottom: 12px; padding-top: 15px">
-                <p>{{ answer.body| safe }}</p>
+                {% if user|is_moderator and MODERATOR_ACTIVATED or user|can_edit:answer and answer|havenot_comments %}
```
Comment (sunilshetye) [inline static/website/templates/get-question.html]: use brackets

File: static/website/templates/get-question.html [hunk 32]
```diff
@@ -265,17 +287,44 @@ <h5>Why is this answer being deleted?</h5>
...
                 <div class="comment comment{{ comment.id }}" style="padding: 15px 15px 15px 5px;">
                     <div class="body" id="cbody{{ comment.id }}">
-                        <p>{{ comment.body| safe }}</p>
+                        {{ comment.body| safe }}
                     </div>
                     
                     <br>
-
-                    {% if user|is_moderator and MODERATOR_ACTIVATED %}
+                    {% if user|is_moderator and MODERATOR_ACTIVATED or user|can_edit:comment and answer|can_delete:comment.id %}
```
Comment (sunilshetye) [inline static/website/templates/get-question.html]: use brackets

File: static/website/templates/moderator/questions.html [hunk 33]
```diff
@@ -12,26 +12,39 @@
...
-		$("#myTable").DataTable({
-          // 'sPaginationType': "scrolling"
+		var table = $("#myTable").DataTable({
           });
+        $('#category').change(function(e){
+            var selectedCategory = $(this).children("option:selected").val();
+            if( selectedCategory != "All Categories")
+                table.column(1).search('^' + selectedCategory + '$', true, false, true).draw();
+            else
+                table.column(1).search('').draw();
```
Comment (sunilshetye) [inline static/website/templates/moderator/questions.html:22]: will **table.column(1).draw()** work?

File: static/website/templates/moderator/unanswered.html [hunk 34]
```diff
@@ -13,9 +13,16 @@ <h5 style="padding-top: 15px;">All unanswered questions</h5>
...
-          $("#myTable").DataTable({
+          var table = $("#myTable").DataTable({
           // 'sPaginationType': "scrolling"
           });
+          $('#category').change(function(e){
+            var selectedCategory = $(this).children("option:selected").val();
+            if( selectedCategory != "All Categories")
+                table.column(1).search('^' + selectedCategory + '$', true, false, true).draw();
+            else
+                table.column(1).search('').draw();
```
Comment (sunilshetye) [inline static/website/templates/moderator/unanswered.html:24]: do not search for empty string.

File: static/website/templates/questions.html [hunk 35]
```diff
@@ -11,9 +11,16 @@
...
-		$("#myTable").DataTable({
+		var table = $("#myTable").DataTable({
           // 'sPaginationType': "scrolling"
           });
+        $('#category').change(function(e){
+            var selectedCategory = $(this).children("option:selected").val();
+            if( selectedCategory != "All Categories")
+                table.column(1).search('^' + selectedCategory + '$', true, false, true).draw();
+            else
+                table.column(1).search('').draw();
```
Comment (sunilshetye) [inline static/website/templates/questions.html:22]: is searching for empty string required?

File: website/cron.py [hunk 36]
```diff
@@ -23,36 +23,36 @@ def unanswered_notification(self):
...
-                uque = Answer.objects.filter(question__id = question.id)
+                uque = Answer.objects.filter(question__id=question.id)
             except Exception as e:
                 print("error occured >  > ")
                 print(e)
             if not uque.exists():
-                i = i+1
+                i = i + 1
                 category_id = question.category.id
                 if category_id in list(body_cat.keys()):
```
Comment (sunilshetye) [inline website/cron.py:43]: will this work?

if category_id in body_cat:

File: website/models.py [hunk 37]
```diff
@@ -5,110 +5,149 @@
...
-    email = models.CharField(max_length = 50)
+    date_created = models.DateTimeField(auto_now_add=True)
+    date_modified = models.DateTimeField(auto_now=True)
+    email = models.CharField(max_length=50)
+    image = ResizedImageField(
+        size=[
+            800,
+            800],
+        upload_to="images/fossCategory/",
+        blank=True)
```
Comment (sunilshetye) [inline website/models.py:23]: add blank line before next def

File: website/templatetags/count_tags.py [hunk 38]
```diff
@@ -30,49 +35,84 @@ def get_range(value, arg = ''):
...
+
+
 register.simple_tag(total_question_count)
 
 # retriving total number of answers
+
+
 def total_answer_count():
-    count = Answer.objects.all().count()
+    count = Answer.objects.all(is_active=True).count()
```
Comment (sunilshetye) [inline website/templatetags/count_tags.py]: does all() take a parameter?

File: website/cron.py [hunk 39]
```diff
@@ -23,36 +23,36 @@ def unanswered_notification(self):
...
         category_count = FossCategory.objects.count()
         i = 0
         body_cat = {}
         for question in questions:
             try:
-                uque = Answer.objects.filter(question__id = question.id)
+                uque = Answer.objects.filter(question__id=question.id)
             except Exception as e:
                 print("error occured >  > ")
                 print(e)
```
Comment (sunilshetye) [inline website/cron.py:39]: there should be a continue in the except block. otherwise uque may not be defined afterwards.

File: website/cron.py [hunk 40]
```diff
@@ -23,36 +23,36 @@ def unanswered_notification(self):
...
             if not uque.exists():
-                i = i+1
+                i = i + 1
```
Comment (sunilshetye) [inline website/cron.py:41]: is **i** being used later?

File: website/views.py [hunk 41]
```diff
@@ -17,27 +17,49 @@
...
+    9, 4376, 4915, 14595, 12329, 22467, 5518, 30705
 )
-categories = FossCategory.objects.order_by('name')
 
 # Function to check if user is in any moderator group
-def is_moderator(user):
+
+
+
+def is_moderator(user, question=None):
```
Comment (sunilshetye) [inline website/views.py:32]: add more documentation for this change

File: website/views.py [hunk 42]
```diff
@@ -46,41 +68,55 @@ def home(request):
...
     }
     return render(request, 'website/templates/questions.html', context)
 
 # get particular question, with votes,anwsers
-def get_question(request, question_id = None, pretty_url = None):
 
-    question = get_object_or_404(Question, id = question_id)
+
+def get_question(request, question_id=None, pretty_url=None):
+    if is_moderator(request.user,get_object_or_404(Question, id=question_id)) and settings.MODERATOR_ACTIVATED:
```
Comment (sunilshetye) [inline website/views.py:89]: use a variable to combine these function calls:

get_object_or_404(Question, id=question_id)
get_object_or_404(Question, id=question_id, is_active=True)

File: website/views.py [hunk 43]
```diff
@@ -670,36 +725,59 @@ def question_delete(request, question_id):
...
-    answer = get_object_or_404(Answer, id = answer_id)
+    answer = get_object_or_404(Answer, id=answer_id)
     question_id = answer.question.id
 
-    if (request.method == "POST"):
+    if ((request.user.id != answer.uid or AnswerComment.objects.filter(answer=answer,
+                                                                       is_active=True).exclude(uid=answer.uid).exists()) and (not is_moderator(request.user, answer.question) or not settings.MODERATOR_ACTIVATED)):
+        return render(request, 'website/templates/not-authorized.html')
+
+    if (request.method == "POST") and (settings.MODERATOR_ACTIVATED):
```
Comment (sunilshetye) [inline website/views.py:774]: is this MODERATOR_ACTIVATED check required again? all authorization checks are already done.

**Review by sunilshetye — COMMENTED** [review]
please check my review also.

commit is too large.

File: static/website/templates/get-question.html (hunk 30 above)
Comment (md-kasim) [reply static/website/templates/get-question.html]: Cann't use brackets in if condition in templates

Comment (sunilshetye) [reply static/website/templates/get-question.html]:
```suggestion
            {% if can_edit_question(user, ans_count) %}
```

in that case, please define a function with appropriate parameters and call that function from here. Suggested code (please fix and use):

```
def can_edit_question(user, ans_count):
    if user|can_edit:question and not ans_count:
        return True
    if user|is_moderator and MODERATOR_ACTIVATED:
        return True
    return False
```

File: static/website/templates/get-question.html (hunk 25 above)
Comment (sunilshetye) [reply static/website/templates/get-question.html]:
```suggestion
            {% if can_edit_answer(user, answer) %}
```

in that case, please define a function with appropriate parameters and call that function from here. Function as per current code (please fix and use):

```
def can_edit_answer(user, answer):
    if user|is_moderator and MODERATOR_ACTIVATED:
        return True
    if user|can_edit:answer and answer|havenot_comments:
        return True
    return False
```

File: static/website/templates/get-question.html [hunk 44]
```diff
@@ -90,9 +96,13 @@
         {% endif %}
 
         <span class="modify">
-            {% if user|can_edit:question and not ans_count or user|is_moderator and MODERATOR_ACTIVATED %}
+            {% if user|can_edit:question and not ans_count or MODERATOR_ACTIVATED %}
```
Comment (sunilshetye) [inline static/website/templates/get-question.html:99]: The logic here is:

```
if user|can_edit:question and not ans_count:
   show the edit link
else if MODERATOR_ACTIVATED:
   show the edit link
```

so ans_count is checked only for the first part. is this logic correct?

File: static/website/templates/get-question.html [hunk 45]
```diff
@@ -149,10 +148,26 @@ <h4><u>Answers:</u></h4>
             </div>
         </div>
 
-        <div class="answer {% if user|is_moderator %}editable{% endif %}" id="answer{{answer.id}}" style="margin-left: 63px;">                               
-            
+        <div class="answer {% if MODERATOR_ACTIVATED or user|can_edit:answer %}editable{% endif %}" id="answer{{answer.id}}" style="margin-left: 63px;">                               
             <div class="body" id="body{{ answer.id }}" style="padding-bottom: 12px; padding-top: 15px">
-                <p>{{ answer.body| safe }}</p>
+                {% if MODERATOR_ACTIVATED or user|can_edit:answer and answer|havenot_comments %}
```
Comment (sunilshetye) [inline static/website/templates/get-question.html:153]: The logic here is:

```
if MODERATOR_ACTIVATED:
   show the form
else if user|can_edit:answer and answer|havenot_comments:
   show the form
```

so havenot_comments is checked only for the second part. is this logic correct?

File: static/website/templates/get-question.html [hunk 46]
```diff
@@ -265,17 +285,44 @@ <h5>Why is this answer being deleted?</h5>
...
                 <div class="comment comment{{ comment.id }}" style="padding: 15px 15px 15px 5px;">
                     <div class="body" id="cbody{{ comment.id }}">
-                        <p>{{ comment.body| safe }}</p>
+                        {{ comment.body| safe }}
                     </div>
                     
                     <br>
-
-                    {% if user|is_moderator and MODERATOR_ACTIVATED %}
+                    {% if MODERATOR_ACTIVATED or user|can_edit:comment and answer|can_delete:comment.id %}
```
Comment (sunilshetye) [inline static/website/templates/get-question.html:295]: The logic here is:

```
if MODERATOR_ACTIVATED:
   show the form
else if user|can_edit:comment and answer|can_delete:comment.id:
   show the form
```

so can_delete is checked only for the second part. is this logic correct?

**Review by sunilshetye — COMMENTED** [review]
please confirm the logic as mentioned in the comments.

File: static/website/templates/get-question.html (hunk 44 above)
Comment (md-kasim) [reply static/website/templates/get-question.html:99]: Yes, this is correct.
not ans_count is being checked only for second part.
Moderator can edit the question even if some answers are posted.
But user can edit only when no answer is being posted

File: static/website/templates/get-question.html (hunk 45 above)
Comment (md-kasim) [reply static/website/templates/get-question.html:153]: Yes, this is correct.

File: static/website/templates/get-question.html (hunk 46 above)
Comment (md-kasim) [reply static/website/templates/get-question.html:295]: Yes, this is correct.

File: static/website/templates/get-question.html (hunk 44 above)
Comment (sunilshetye) [reply static/website/templates/get-question.html:99]: okay, thanks for all the confirmations.

File: forums/settings.py (hunk 28 above)
Comment (prashantsinalkar) [reply forums/settings.py:167]: This is the security related setting. User session expires at as soon as browser closed by the user. He needs to login again.

File: static/website/js/thread-user.js (hunk 24 above)
Comment (prashantsinalkar) [reply static/website/js/thread-user.js:13]: Its for the edit and the ck-editor wrapper for the edit answer for multiple answers.

File: website/cron.py (hunk 40 above)
Comment (prashantsinalkar) [reply website/cron.py:41]: Currently its not used any where. Need to rewrite this code. As its existing code.

File: website/cron.py (hunk 4 above)
Comment (prashantsinalkar) [reply website/cron.py:29]: This is the cron file. Its independent code and the changes shown are just pep8 fixes. We will figure out later.

File: website/cron.py (hunk 39 above)
Comment (prashantsinalkar) [reply website/cron.py:39]: Currently its not used any where. Need to rewrite this code. As its existing code. We will fix it later as its a independent code.

File: static/website/js/thread-user.js (hunk 24 above)
Comment (md-kasim) [reply static/website/js/thread-user.js:13]: its for niceditor.

File: forums/settings.py (hunk 28 above)
Comment (md-kasim) [reply forums/settings.py:167]: I should remove this line because this is a forum site. So Security should not be concerned too much level.

File: website/cron.py (hunk 40 above)
Comment (md-kasim) [reply website/cron.py:41]: has been rewritten

File: website/cron.py (hunk 39 above)
Comment (md-kasim) [reply website/cron.py:39]: has been rewritten

File: website/views.py (hunk 43 above)
Comment (prashantsinalkar) [reply website/views.py:774]: Have you checked this one?

Comment (md-kasim) [reply website/views.py:774]: this is required because mail is sent only when moderator deletes the answer.
So When authorised user deletes the answer, mail is not sent to anyone.
User can only be authorized to delete an answer if he has posted this answer and no comment is being posted under this answer.

File: static/website/templates/index.html [hunk 47]
```diff
@@ -3,102 +3,113 @@
...
+    <div class="panel-heading">
+        Recent questions
+    </div>
+
+    <div class="panel-body">
+
+        <table id="myTable" class="tablesorter-blue">
+            Category : 
+            <select name="categories" id="category" style="margin: 10px 0px 15px 5px;">
+                <option>All Categories</option>
```
Comment (prashantsinalkar) [inline static/website/templates/index.html:82]: Can not render the sub category as well as questions list for scilab-toolbox.

Comment (md-kasim) [reply static/website/templates/index.html:82]: solved
